### ISSUE TITLE
Failed to handle front-end request

### ISSUE THREAD

**Issue by gacou54** [FAIRsharing/FAIRsharing-API#136]
We would like to use FAIRSharing at our institution (having our own instance to list tools and databases). We have some problems when we try to launch the application with the front end https://github.com/FAIRsharing/fairsharing.github.io.

To try the application in developer mode, I launched the back-end and the front-end on my machine. In the front end, I changed the variable `this.url` in the files` src/components/Client/RESTClient.js` and
`src/components/GraphClient/GraphClient.js` to point to the local instance.

When I try to open the web page, the back-end throws the following error
```
Started OPTIONS "/graphql" for 127.0.0.1 at 2020-05-05 12:06:06 -0400
   (19.0ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC
Rack attack: safelist.rack_attack, 2020-05-05 12:06:07 -0400, 2020-05-05 12:06:07 -0400, ee2c8cc7e9e25158f114,
                                   ,

ActionController :: RoutingError (No route matches [OPTIONS] "/ graphql"):

  
actionpack (6.0.2.2) lib / action_dispatch / middleware / debug_exceptions.rb: 36: in `call '
web-console (4.0.1) lib / web_console / middleware.rb: 132: in `call_app '
web-console (4.0.1) lib / web_console / middleware.rb: 28: in `block in call '
web-console (4.0.1) lib / web_console / middleware.rb: 17: in `catch '
web-console (4.0.1) lib / web_console / middleware.rb: 17: in `call '
actionpack (6.0.2.2) lib / action_dispatch / middleware / show_exceptions.rb: 33: in `call '
railties (6.0.2.2) lib / rails / rack / logger.rb: 38: in `call_app '
railties (6.0.2.2) lib / rails / rack / logger.rb: 26: in `block in call '
activesupport (6.0.2.2) lib / active_support / tagged_logging.rb: 80: in `block in tagged '
activesupport (6.0.2.2) lib / active_support / tagged_logging.rb: 28: in `tagged '
activesupport (6.0.2.2) lib / active_support / tagged_logging.rb: 80: in `tagged '
railties (6.0.2.2) lib / rails / rack / logger.rb: 26: in `call '
sprockets-rails (3.2.1) lib / sprockets / rails / quiet_assets.rb: 13: in `call '
actionpack (6.0.2.2) lib / action_dispatch / middleware / remote_ip.rb: 81: in `call '
request_store (1.5.0) lib / request_store / middleware.rb: 19: in `call '
actionpack (6.0.2.2) lib / action_dispatch / middleware / request_id.rb: 27: in `call '
rack (2.2.2) lib / rack / runtime.rb: 22: in `call '
activesupport (6.0.2.2) lib / active_support / cache / strategy / local_cache_middleware.rb: 29: in `call '
actionpack (6.0.2.2) lib / action_dispatch / middleware / executor.rb: 14: in `call '
actionpack (6.0.2.2) lib / action_dispatch / middleware / static.rb: 126: in `call '
rack (2.2.2) lib / rack / sendfile.rb: 110: in `call '
actionpack (6.0.2.2) lib / action_dispatch / middleware / host_authorization.rb: 83: in `call '
railties (6.0.2.2) lib / rails / engine.rb: 526: in `call '
puma (4.3.3) lib / puma / configuration.rb: 228: in `call '
puma (4.3.3) lib / puma / server.rb: 682: in `handle_request '
puma (4.3.3) lib / puma / server.rb: 472: in `process_client '
puma (4.3.3) lib / puma / server.rb: 328: in `block in run '
puma (4.3.3) lib / puma / thread_pool.rb: 134: in `block in spawn_thread '
```

Do you have an idea of ​​the problem, or did I misunderstand something?

Also, I think I saw an issue where a docker container is discussed. This could greatly facilitate deployment for people without experience with the application and ruby ​​on rails.

Thank you very much,
Gabriel

**Comment by smichi23** [comment]
I had the same problem. I could partially solve the problem by modifying the FAIRsharing-API/config/application.rb file.
`` 
require_relative 'boot'

require 'rails/all'

# Require the gems listed in Gemfile, including any gems
# you've limited to :test, :development, or :production.
Bundler.require(*Rails.groups)

module FairsharingApi
  class Application < Rails::Application
    # Initialize configuration defaults for originally generated Rails version.
    config.api_only = true
    config.load_defaults 6.0
    config.autoload_paths += %W(#{config.root}/lib/modules)
    config.action_dispatch.return_only_media_type_on_content_type = false
    config.active_job.queue_adapter = :queue_classic

    # Settings in config/environments/* take precedence over those specified here.
    # Application configuration can go into files in config/initializers
    # -- all .rb files in that directory are automatically loaded after loading
    # the framework and any gems in your application.

    # This has been disabled due to an extra Access-Control-Allow-Origin header being sent,
    # to the dismay of the front-end javascript application.
    #config.middleware.insert_before 0, Rack::Cors do
    #  allow do
    #    origins '*'
    #    resource '*', headers: :any, methods: [:get, :post, :options]
    #  end
    #end

    # Workaround for Omniauth
    config.middleware.use Rack::Session::Cookie, secret: Rails.application.credentials.session_cookie_secret

  end
end

**Comment by smichi23** [comment]
I had the same problem, I could partially solve it by modifying the FAIRsharing-API/config/application.rb file
```
  
require_relative 'boot'

require 'rails/all'

# Require the gems listed in Gemfile, including any gems
# you've limited to :test, :development, or :production.
Bundler.require(*Rails.groups)

module FairsharingApi
  class Application < Rails::Application
    # Initialize configuration defaults for originally generated Rails version.
    config.api_only = true
    config.load_defaults 6.0
    config.autoload_paths += %W(#{config.root}/lib/modules)
    config.action_dispatch.return_only_media_type_on_content_type = false
    config.active_job.queue_adapter = :queue_classic

    # Settings in config/environments/* take precedence over those specified here.
    # Application configuration can go into files in config/initializers
    # -- all .rb files in that directory are automatically loaded after loading
    # the framework and any gems in your application.

    # This has been disabled due to an extra Access-Control-Allow-Origin header being sent,
    # to the dismay of the front-end javascript application.
    config.middleware.insert_before 0, Rack::Cors do
      allow do
        origins 'http://localhost:8080'
        resource '*', headers: :any, methods: [:get, :post, :options]
      end
    end

    # Workaround for Omniauth
    config.middleware.use Rack::Session::Cookie, secret: Rails.application.credentials.session_cookie_secret

  end
end
```
Doing this solved the preflights and the queries problems. Even when my queries got through to the backend they didn't seem to work. When I open the frontend, I get in the backend logs 
```
Processing by GraphqlController#execute as JSON
  Parameters: {"query"=>" { __schema {   types  {     ...FullType   } }  }  fragment FullType on __Type { name fields(includeDeprecated: true) {   name   description   args {     ...InputValue   }   isDeprecated }  }  fragment InputValue on __InputValue { name description type { ...TypeRef } defaultValue  }  fragment TypeRef on __Type { kind name ofType {   kind   name   ofType {     kind     name   } }  }", "graphql"=>{"query"=>" { __schema {   types  {     ...FullType   } }  }  fragment FullType on __Type { name fields(includeDeprecated: true) {   name   description   args {     ...InputValue   }   isDeprecated }  }  fragment InputValue on __InputValue { name description type { ...TypeRef } defaultValue  }  fragment TypeRef on __Type { kind name ofType {   kind   name   ofType {     kind     name   } }  }"}}
[active_model_serializers] Rendered ActiveModel::Serializer::Null with Hash (0.08ms)
Completed 501 Not Implemented in 1ms (Views: 0.7ms | ActiveRecord: 0.0ms | Allocations: 355)
```
If I copy paste the graphql query in the graphiql at localhost:3000 (port of the back end) I
almost get the exact same logs but the query works.
```
Processing by GraphqlController#execute as */*
  Parameters: {"query"=>" { __schema {   types  {     ...FullType   } }  }  fragment FullType on __Type { name fields(includeDeprecated: true) {   name   description   args {     ...InputValue   }   isDeprecated }  }  fragment InputValue on __InputValue { name description type { ...TypeRef } defaultValue  }  fragment TypeRef on __Type { kind name ofType {   kind   name   ofType {     kind     name   } }  }", "variables"=>nil, "graphql"=>{"query"=>" { __schema {   types  {     ...FullType   } }  }  fragment FullType on __Type { name fields(includeDeprecated: true) {   name   description   args {     ...InputValue   }   isDeprecated }  }  fragment InputValue on __InputValue { name description type { ...TypeRef } defaultValue  }  fragment TypeRef on __Type { kind name ofType {   kind   name   ofType {     kind     name   } }  }", "variables"=>nil}}
[active_model_serializers] Rendered ActiveModel::Serializer::Null with Hash (10.16ms)
Completed 200 OK in 71ms (Views: 11.1ms | ActiveRecord: 0.0ms | Allocations: 45333)
```
If you have any tips or informations that could help me fix that problem, they would be appreciated.

**Comment by knirirr** [comment]
The CORS headers settings are needed for the application to run, but they have been commented out in the code as when running in production the same options can be set in the webserver configuration - see https://geekflare.com/enable-cors-apache-nginx/ for examples. Both options can't be used simultaneously.  

`ActionController :: RoutingError (No route matches [OPTIONS] "/ graphql"):` - there's a space in the path which is presumably causing the 404. 

The `ActiveModel::Serializer::Null` is presumably as there's no data in the database, which you would have to populate (there are examples in the fixtures). 

Please note that the FAIRsharing code is still in pre-alpha stage and under heavy development, so it's not really suitable for production use at present; even the licence terms are still under discussion. I'm afraid that Docker is of low priority and will probably be investigated once we have something running here.
